rust/kernel/str.rs

```diff
@@ -247,3 +252,187 @@ macro_rules! c_str {
+        build_assert!(N <= M, "relaxed bound should be no less than current bound");
+        unsafe { CBoundedStr::<M>::from_c_str_unchecked(&self.0) }
+    }
+
+    /// Converts the string to a c_char array with the current bound, fills the
+    /// remaining bytes with zero.
+    #[inline]
+    pub const fn to_char_array(&self) -> [c_types::c_char; N] {
+        let mut ret: [c_types::c_char; N] = [0; N];
+        let mut i = 0;
+        while i < self.0 .0.len() {
+            ret[i] = self.0 .0[i] as _;
+            i += 1;
+        }
+        ret
+    }
+
+    /// Expands the string a c_char array, fills remaining bytes with zero.
+    ///
+    /// `M` must be no less than the bound `N`.
+    #[inline]
+    pub const fn expand_to_char_array<const M: usize>(&self) -> [c_types::c_char; M] {
+        self.relax_bound().to_char_array()
+    }
+}
+
+impl<const N: usize> AsRef<BStr> for CBoundedStr<N> {
+    #[inline]
+    fn as_ref(&self) -> &BStr {
+        self.as_bytes()
+    }
+}
+
+impl<const N: usize> AsRef<CStr> for CBoundedStr<N> {
+    #[inline]
+    fn as_ref(&self) -> &CStr {
+        &self.0
+    }
+}
+
+impl<const N: usize> Deref for CBoundedStr<N> {
+    type Target = CStr;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<Idx, const N: usize> Index<Idx> for CBoundedStr<N>
+where
+    CStr: Index<Idx>,
+{
+    type Output = <CStr as Index<Idx>>::Output;
+
+    #[inline]
+    fn index(&self, index: Idx) -> &Self::Output {
+        &self.0[index]
+    }
+}
+
+/// Creates a new [`CBoundedStr`] from a string literal.
+///
+/// The string literal should not contain any `NUL` bytes, and its length with NUL should not
+/// exceed the bound supplied.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// // If no bound is specified, the tighest bound will be inferred:
+/// const MY_CSTR: &'static CBoundedStr<17> = c_bounded_str!("My awesome CStr!");
+/// ```
+///
+/// ```rust,compile_fail
+/// // This would not compile as the type is `CBoundedStr<17>`.
+/// const MY_CSTR: &'static CBoundedStr<100> = c_bounded_str!("My awesome CStr!");
+/// ```
+///
+/// ```rust,no_run
+/// // You can relax the bound using the `relax_bound` method.
+/// const MY_CSTR: &'static CBoundedStr<100> = c_bounded_str!("My awesome CStr!").relax_bound();
+///
+/// // Or alternatively specify a bound when invoking macro.
+/// // In this case the supplied bound must be a constant expression.
+/// const MY_CSTR2: &'static CBoundedStr<100> = c_bounded_str!(100, "My awesome CStr!");
+///
+/// // Or let the compiler infer the bound for you.
+/// const MY_CSTR3: &'static CBoundedStr<100> = c_bounded_str!(_, "My awesome CStr!");
+/// ```
+///
+/// ```rust,compile_fail
+/// // shouldn't compile as the string is longer than the specified bound.
+/// const MY_CSTR: &'static CBoundedStr<10> = c_bounded_str!(10, "My awesome CStr!");
+/// const MY_CSTR2: &'static CBoundedStr<10> = c_bounded_str!(_, "My awesome CStr!");
```

Comment:
Perhaps to be more clear:

```suggestion
/// const MY_CSTR: &'static CBoundedStr<4> = c_bounded_str!(4, "My awesome CStr!");
/// const MY_CSTR2: &'static CBoundedStr<4> = c_bounded_str!(_, "My awesome CStr!");
```